docs/proposals/L7-traffic-route.md

```diff
@@ -0,0 +1,174 @@
+# L7 Traffic Routing
+
+## Context
+
+Kuma provides traffic management with `TrafficRoute` policy based on Kuma tags.
+We can split the traffic to different destinations, we can redirect the traffic etc.
+
+## Requirements
+
+Provide a way to manage the traffic based on HTTP headers, path and method.
+
+## Configuration
+
+```yaml
+type: TrafficRoute
+name: route-all-default
+mesh: default
+sources:
+- match:
+    kuma.io/service: web
+destinations:
+- match:
+    kuma.io/service: backend
+conf:
+  http: # a new http section which is a list
+  - match: # match is required and need to have at least one sub-element
+      method: GET
+      path: # one of either prefix, exact or regex will be allowed
+        prefix: /users
+        exact: /users/user-1
+        regex: 
```

Comment:
Suggest a match-type field:
```
path:
    type: exact
    value: /users/
```

This is consistent with Kubernetes API conventions and allows valid match types to be automatically validated by API annotations. It also makes validation check unambigious.

Reply:
However, there is a feature of protobuf called "oneof" and providing a couple of options and picking one of them is a convention of Envoy. We use this already in many places. Even `loadBalancer` in TrafficRoute

Reply:
Were they also migrating from "oneof" convention?
Validation on API annotations? I don't think I'm familiar with this, can you tell me more? We have our own unified validators for both Kubernetes and Universal.

Reply:
The kubernetes CRD generator support validation that will propagate checks into the openapi spec (see kubebuilder docs).

I'm confide about the protobuf reference. Is this a protobuf API? If so, shouldn't it be in proto3 rather than in yaml?

Reply:
Kuma API is described in protobuf and then translated between JSON / YAML (but also sent between control planes in multizone deployment as serialized protobuf). In proposals, we tend to use YAML examples so it's easier to read, but you might be right that maybe we should also include proto schemas.

Protobuf also has a builtin validation functionality but we found this to not be flexible enough (descriptive messages) for our needs

Reply:
Ah I see. I'd probably design a YAML-oriented API differently than a protobuf-oriented API :)

Reply:
FWIW, Kubernetes CRD validation doesn't support unions, or one-of. There are generator annotations for it, but they don't do anything :joy: